x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java

```diff
@@ -68,7 +79,19 @@ private DeprecationIndexingComponent(
         this.clusterService = clusterService;
 
         this.processor = getBulkProcessor(new OriginSettingClient(client, ClientHelper.DEPRECATION_ORIGIN), settings);
-        final Consumer<IndexRequest> consumer = this.processor::add;
+        final Consumer<IndexRequest> consumer = request -> {
+            /*
+             * It is possible that deprecationLoggingEnabled is false in the first check but then becomes true (and the queue is
+             * drained) before the call to offer. Given the rarity of this case, and the low impact of dropping a single deprecation log
+             * message, we're living with this race condition instead of impacting performance (adding a mutex) or complexity and
+             * performance (adding a periodic task to check the queue).
+             */
+            if (flushEnabled.get()) {
+                processor.add(request);
```

Comment:
Had an idea about this. What if we drain the queue in this location as well? It shouldn't replay any previously removed items, and it will clear any items that might have snuck into the queue since the drain operation in the cluster management thread.

Might be overkill, but this is a clunky thing already without fully implementing a producer consumer model.

Reply:
That's a good idea. Thanks.